src/passes/MemoryCopyFillLowering.cpp

```diff
@@ -0,0 +1,184 @@
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+// Replace memory.copy and memory.fill with a call to a function that
+// implements the same semantics.
+
+namespace wasm {
+struct MemoryCopyFillLowering
+  : public WalkerPass<PostWalker<MemoryCopyFillLowering>> {
+  bool needsMemoryCopy = false;
+  bool needsMemoryFill = false;
+
+  void visitMemoryCopy(MemoryCopy* curr) {
+    if (curr->destMemory != curr->sourceMemory) {
+      return; // Throw an error here instead of silently ignoring?
+    }
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeCall(
+        "__memory_copy", {curr->dest, curr->source, curr->size},
+        Type::none));
+    needsMemoryCopy = true;
+  }
+
+  void visitMemoryFill(MemoryFill* curr) {
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeCall(
+        "__memory_fill", {curr->dest, curr->value, curr->size}, Type::none)
+    );
+    needsMemoryFill = true;
+  }
+
+  void run(Module* module) override {
+    if (!module->features.hasBulkMemory()) {
+      return;
+    }
+    if (module->features.hasMemory64() || module->features.hasMultiMemory()) {
+      throw "Memory64 and multi-memory not supported";// TODO: best way to return an error?
+    }
+    // In order to introduce a call to a function, it must first exist, so create an empty stub.
+    Builder b(*module);
+    auto memCpyFunc = b.makeFunction("__memory_copy",
+      Signature({Type::i32, Type::i32, Type::i32}, {Type::none}), {Type::i32});
+    memCpyFunc->body = b.makeBlock();
+    module->addFunction(memCpyFunc.release());
+    auto memFillFunc = b.makeFunction("__memory_fill",
+      Signature({Type::i32, Type::i32, Type::i32}, {Type::none}), {});
+    memFillFunc->body = b.makeBlock();
+    module->addFunction(memFillFunc.release());
+
+    Super::run(module);
+
+    if (needsMemoryCopy) {
+      Index dst = 0, src = 1, size = 2, temp = 3;
+      Name memory = module->memories.front()->name;
```

Comment:
It's worse than that, because there can also be cross-memory copies. And one or both memories could be 64-bit. There's a check up top that bails if the module has multi-memory or memory64 support.

Reply:
Heh, right, cross-memory copies too... sgtm to limit to single memories, I missed that part.